### PR TITLE
JDK-8301072: Replace NULL with nullptr in share/oops/

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi, this PR changes all occurrences of NULL to nullptr for the subdirectory share/oops/. Unfortunately the script that does the change isn't perfect, and so we
need to comb through these manually to make sure nothing has gone wrong. I also review these changes but things slip past my eyes sometimes.

Here are some typical things to look out for:

1. No changes but copyright header changed (probably because I reverted some changes but forgot the copyright).
2. Macros having their NULL changed to nullptr, these are added to the script when I find them. They should be NULL.
3. nullptr in comments and logs. We try to use lower case "null" in these cases as it reads better. An exception is made when code expressions are in a comment.

An example of this:

```c++
// This function returns null
void* ret_null();
// This function returns true if *x == nullptr
bool is_nullptr(void** x);
```

Note how `nullptr` participates in a code expression here, we really are talking about the specific value `nullptr`.

Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301072](https://bugs.openjdk.org/browse/JDK-8301072): Replace NULL with nullptr in share/oops/

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to [9d9fdfc7](https://git.openjdk.org/jdk/pull/12186/files/9d9fdfc76ba43cde6d9ab8c76e91fe9a5207f628)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [9d9fdfc7](https://git.openjdk.org/jdk/pull/12186/files/9d9fdfc76ba43cde6d9ab8c76e91fe9a5207f628)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [f6f0d873](https://git.openjdk.org/jdk/pull/12186/files/f6f0d873ce54f0aecff5e9dc49c9fbd168340dda)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12186/head:pull/12186` \
`$ git checkout pull/12186`

Update a local copy of the PR: \
`$ git checkout pull/12186` \
`$ git pull https://git.openjdk.org/jdk pull/12186/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12186`

View PR using the GUI difftool: \
`$ git pr show -t 12186`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12186.diff">https://git.openjdk.org/jdk/pull/12186.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/oops/annotations.cpp [hunk 1]
```diff
@@ -81,7 +81,7 @@ void Annotations::metaspace_pointers_do(MetaspaceClosure* it) {
   it->push(&_class_annotations);
   it->push(&_fields_annotations);
   it->push(&_class_type_annotations);
-  it->push(&_fields_type_annotations); // FIXME: need a test case where _fields_type_annotations != NULL
+  it->push(&_fields_type_annotations); // FIXME: need a test case where _fields_type_annotations != null
```
Comment (jdksjolen) [inline src/hotspot/share/oops/annotations.cpp:84]: nullptr

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 2]
```diff
@@ -111,10 +111,10 @@
 #define HOTSPOT_CLASS_INITIALIZATION_end HOTSPOT_CLASS_INITIALIZATION_END
 #define DTRACE_CLASSINIT_PROBE(type, thread_type)                \
   {                                                              \
-    char* data = NULL;                                           \
+    char* data = nullptr;                                           \
```
Comment (jdksjolen) [inline src/hotspot/share/oops/instanceKlass.cpp:114]: align

File: src/hotspot/share/oops/method.cpp [hunk 3]
```diff
@@ -1614,12 +1614,12 @@ methodHandle Method::clone_with_new_data(const methodHandle& m, u_char* new_code
 }
 
 vmSymbolID Method::klass_id_for_intrinsics(const Klass* holder) {
-  // if loader is not the default loader (i.e., != NULL), we can't know the intrinsics
+  // if loader is not the default loader (i.e., != null), we can't know the intrinsics
```
Comment (jdksjolen) [inline src/hotspot/share/oops/method.cpp:1617]: non-null

**Review by jdksjolen — COMMENTED** [review]
Found some manual fixes

**Comment by jdksjolen** [comment]
Passes tier1.

File: src/hotspot/share/oops/compressedOops.inline.hpp [hunk 4]
```diff
@@ -60,7 +60,7 @@ inline oop CompressedOops::decode_not_null(narrowOop v) {
 }
 
 inline oop CompressedOops::decode(narrowOop v) {
-  return is_null(v) ? (oop)NULL : decode_not_null(v);
+  return is_null(v) ? (oop)nullptr : decode_not_null(v);
```
Comment (stefank) [inline src/hotspot/share/oops/compressedOops.inline.hpp:63]: Seems like we should be able to remove the `(oop)` cast after this change.

File: src/hotspot/share/oops/compressedOops.inline.hpp [hunk 5]
```diff
@@ -137,7 +137,7 @@ inline Klass* CompressedKlassPointers::decode_not_null(narrowKlass v, address na
 }
 
 inline Klass* CompressedKlassPointers::decode(narrowKlass v) {
-  return is_null(v) ? (Klass*)NULL : decode_not_null(v);
+  return is_null(v) ? (Klass*)nullptr : decode_not_null(v);
```
Comment (stefank) [inline src/hotspot/share/oops/compressedOops.inline.hpp:140]: Seems like we should be able to remove the `(Klass*)` cast after this change.

File: src/hotspot/share/oops/constantPool.hpp [hunk 6]
```diff
@@ -186,15 +186,15 @@ class ConstantPool : public Metadata {
   // generics support
   Symbol* generic_signature() const {
     return (_generic_signature_index == 0) ?
-      (Symbol*)NULL : symbol_at(_generic_signature_index);
+      (Symbol*)nullptr : symbol_at(_generic_signature_index);
```
Comment (stefank) [inline src/hotspot/share/oops/constantPool.hpp:189]: Remove cast?

File: src/hotspot/share/oops/constantPool.hpp [hunk 7]
```diff
@@ -186,15 +186,15 @@ class ConstantPool : public Metadata {
...
   }
   u2 generic_signature_index() const                   { return _generic_signature_index; }
   void set_generic_signature_index(u2 sig_index)       { _generic_signature_index = sig_index; }
 
   // source file name
   Symbol* source_file_name() const {
     return (_source_file_name_index == 0) ?
-      (Symbol*)NULL : symbol_at(_source_file_name_index);
+      (Symbol*)nullptr : symbol_at(_source_file_name_index);
```
Comment (stefank) [inline src/hotspot/share/oops/constantPool.hpp:197]: Remove cast?

File: src/hotspot/share/oops/generateOopMap.cpp [hunk 8]
```diff
@@ -674,7 +674,7 @@ int GenerateOopMap::next_bb_start_pc(BasicBlock *bb) {
 // Allocate memory and throw LinkageError if failure.
 #define ALLOC_RESOURCE_ARRAY(var, type, count) \
   var = NEW_RESOURCE_ARRAY_RETURN_NULL(type, count);              \
-  if (var == NULL) {                                              \
+  if (var == nullptr) {                                              \
```
Comment (stefank) [inline src/hotspot/share/oops/generateOopMap.cpp:677]: Could you fix the alignment of the backslashes?

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 9]
```diff
@@ -2161,20 +2161,20 @@ jmethodID InstanceKlass::get_jmethod_id(const methodHandle& method_h) {
   // implied else:
   // we need to allocate a cache so default length and id values are good
 
-  if (jmeths == NULL ||   // no cache yet
+  if (jmeths == nullptr ||   // no cache yet
       length <= idnum ||  // cache is too short
-      id == NULL) {       // cache doesn't contain entry
+      id == nullptr) {       // cache doesn't contain entry
```
Comment (stefank) [inline src/hotspot/share/oops/instanceKlass.cpp:2166]: Comments seems to be misaligned now.

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 10]
```diff
@@ -2245,19 +2245,19 @@ jmethodID InstanceKlass::get_jmethod_id_fetch_or_update(
...
 
   // reacquire the cache - we are locked, single threaded or at a safepoint
   jmethodID* jmeths = methods_jmethod_ids_acquire();
-  jmethodID  id     = NULL;
+  jmethodID  id     = nullptr;
   size_t     length = 0;
 
-  if (jmeths == NULL ||                         // no cache yet
+  if (jmeths == nullptr ||                         // no cache yet
       (length = (size_t)jmeths[0]) <= idnum) {  // cache is too short
```
Comment (stefank) [inline src/hotspot/share/oops/instanceKlass.cpp:2259]: Alignment of comments

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 11]
```diff
@@ -2293,29 +2293,29 @@ jmethodID InstanceKlass::get_jmethod_id_fetch_or_update(
...
 jmethodID InstanceKlass::jmethod_id_or_null(Method* method) {
   size_t idnum = (size_t)method->method_idnum();
   jmethodID* jmeths = methods_jmethod_ids_acquire();
   size_t length;                                // length assigned as debugging crumb
-  jmethodID id = NULL;
-  if (jmeths != NULL &&                         // If there is a cache
+  jmethodID id = nullptr;
+  if (jmeths != nullptr &&                         // If there is a cache
       (length = (size_t)jmeths[0]) > idnum) {   // and if it is long enough,
-    id = jmeths[idnum+1];                       // Look up the id (may be NULL)
```
Comment (stefank) [inline src/hotspot/share/oops/instanceKlass.cpp:2318]: Alignment of comments

File: src/hotspot/share/oops/klass.cpp [hunk 12]
```diff
@@ -428,17 +428,17 @@ void Klass::append_to_sibling_list() {
   debug_only(verify();)
   // add ourselves to superklass' subklass list
   InstanceKlass* super = superklass();
-  if (super == NULL) return;        // special case: class Object
+  if (super == nullptr) return;        // special case: class Object
   assert((!super->is_interface()    // interfaces cannot be supers
-          && (super->superklass() == NULL || !is_interface())),
```
Comment (stefank) [inline src/hotspot/share/oops/klass.cpp:433]: Alignment of comments

File: src/hotspot/share/oops/method.hpp [hunk 13]
```diff
@@ -164,7 +164,7 @@ class Method : public Metadata {
   void set_signature_index(int index)            { constMethod()->set_signature_index(index);       }
 
   // generics support
-  Symbol* generic_signature() const              { int idx = generic_signature_index(); return ((idx != 0) ? constants()->symbol_at(idx) : (Symbol*)NULL); }
+  Symbol* generic_signature() const              { int idx = generic_signature_index(); return ((idx != 0) ? constants()->symbol_at(idx) : (Symbol*)nullptr); }
```
Comment (stefank) [inline src/hotspot/share/oops/method.hpp:167]: Remove cast?

File: src/hotspot/share/oops/oopHandle.inline.hpp [hunk 14]
```diff
@@ -31,33 +31,33 @@
 #include "gc/shared/oopStorage.inline.hpp"
 
 inline oop OopHandle::resolve() const {
-  return (_obj == NULL) ? (oop)NULL : NativeAccess<>::oop_load(_obj);
+  return (_obj == nullptr) ? (oop)nullptr : NativeAccess<>::oop_load(_obj);
```
Comment (stefank) [inline src/hotspot/share/oops/oopHandle.inline.hpp:34]: Remove cast?

File: src/hotspot/share/oops/oopHandle.inline.hpp [hunk 15]
```diff
@@ -31,33 +31,33 @@
...
 }
 
 inline oop OopHandle::peek() const {
-  return (_obj == NULL) ? (oop)NULL : NativeAccess<AS_NO_KEEPALIVE>::oop_load(_obj);
+  return (_obj == nullptr) ? (oop)nullptr : NativeAccess<AS_NO_KEEPALIVE>::oop_load(_obj);
```
Comment (stefank) [inline src/hotspot/share/oops/oopHandle.inline.hpp:38]: Remove cast?

File: src/hotspot/share/oops/oopHandle.inline.hpp [hunk 16]
```diff
@@ -31,33 +31,33 @@
...
   }
   NativeAccess<>::oop_store(_obj, obj);
 }
 
 inline void OopHandle::release(OopStorage* storage) {
-  if (_obj != NULL) {
+  if (_obj != nullptr) {
     // Clear the OopHandle first
-    NativeAccess<>::oop_store(_obj, (oop)NULL);
+    NativeAccess<>::oop_store(_obj, (oop)nullptr);
```
Comment (stefank) [inline src/hotspot/share/oops/oopHandle.inline.hpp:53]: Remove cast?

File: src/hotspot/share/oops/weakHandle.cpp [hunk 17]
```diff
@@ -48,10 +48,10 @@ WeakHandle::WeakHandle(OopStorage* storage, oop obj) :
 
 void WeakHandle::release(OopStorage* storage) const {
   // Only release if the pointer to the object has been created.
-  if (_obj != NULL) {
+  if (_obj != nullptr) {
     // Clear the WeakHandle.  For race in creating ClassLoaderData, we can release this
     // WeakHandle before it is cleared by GC.
-    NativeAccess<ON_PHANTOM_OOP_REF>::oop_store(_obj, (oop)NULL);
+    NativeAccess<ON_PHANTOM_OOP_REF>::oop_store(_obj, (oop)nullptr);
```
Comment (stefank) [inline src/hotspot/share/oops/weakHandle.cpp:54]: Remove cast?

**Review by stefank — CHANGES_REQUESTED** [review]

**Comment by jdksjolen** [comment]
Thanks Stefan, added your fixes.

**Review by coleenp — APPROVED** [review]
Besides the casts that you may or may not be able to remove, this looks fine.

**Review by stefank — APPROVED** [review]

File: src/hotspot/share/oops/oop.hpp [hunk 18]
```diff
@@ -260,7 +260,7 @@ class oopDesc {
 
   // Like "forward_to", but inserts the forwarding pointer atomically.
   // Exactly one thread succeeds in inserting the forwarding pointer, and
-  // this call returns "NULL" for that thread; any other thread has the
+  // this call returns "null" for that thread; any other thread has the
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/oop.hpp:263]: Nit: null should not be in quotes

**Review by dholmes-ora — APPROVED** [review]
Looks good. Thanks!

**Comment by jdksjolen** [comment]
@stefank , we do need to those `oop` casts. The `NativeAccess<>::oop_store` calls makes sense, `oop_store` is templated so we need to provide the type. We also need the casts in `CompressedOops`, and I don't know why that is.

**Review by dholmes-ora — APPROVED** [review]

**Comment by stefank** [comment]
> @stefank , we do need to those `oop` casts. The `NativeAccess<>::oop_store` calls makes sense, `oop_store` is templated so we need to provide the type. We also need the casts in `CompressedOops`, and I don't know why that is.

I took a look at this. And as far as I can see, there's no obvious problem with the oop_store ore CompressedOops changes, but with the NativeAccess<>::oop_load changes.

NativeAccess<>::oop_load returns an OopLoadProxy object instead of oop, which causes the following code:
```
inline oop OopHandle::resolve() const {
  return (_obj == nullptr) ? nullptr : NativeAccess<>::oop_load(_obj);
}
```
To use the raw `nullptr` and create an OopLoadProxy, which then crashes because the passed in value is null.

We might want to reconsider removing OopLoadProxy, since it isn't the first time it is causing problems for us. I'll think about that in a context of a separate RFE.

**Comment by jdksjolen** [comment]
> > @stefank , we do need to those `oop` casts. The `NativeAccess<>::oop_store` calls makes sense, `oop_store` is templated so we need to provide the type. We also need the casts in `CompressedOops`, and I don't know why that is.
> 
> I took a look at this. And as far as I can see, there's no obvious problem with the oop_store ore CompressedOops changes, but with the NativeAccess<>::oop_load changes.
> 
> NativeAccess<>::oop_load returns an OopLoadProxy object instead of oop, which causes the following code:
> 
> ```
> inline oop OopHandle::resolve() const {
>   return (_obj == nullptr) ? nullptr : NativeAccess<>::oop_load(_obj);
> }
> ```
> 
> To use the raw `nullptr` and create an OopLoadProxy, which then crashes because the passed in value is null.
> 
> We might want to reconsider removing OopLoadProxy, since it isn't the first time it is causing problems for us. I'll think about that in a context of a separate RFE.

Thanks for looking at these. I'm removing the unnecessary casts and keeping the remaining ones.

**Comment by jdksjolen** [comment]
Last GHA passed, integrating this as the merge conflict was trivial.

/integrate

**Pull request closed without merging**